### PR TITLE
[storage] expose state db sizing info to Move

### PR DESCRIPTION
**Pull request by msmouse**
### Description

(the two commits doesn't separate logic cleanly, will need to squash after reviewing)

### Test Plan

strict consistency check on state snapshot commit, so leveraging all integration tests to check SMT::usage is consistent with StateStore::get_usage() which is unit tested.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3021)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/state_storage.move [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            error::already_exists(EUSAGE_ALREADY_EXISTS)
+        );
+        move_to(account, StateStorageUsage {
+            items: 0,
+            bytes: 0,
+        });
+    }
+
+    public(friend) fun on_epoch_begin() {
+        let cfg = borrow_global<Configuration>(@aptos_framework);
```
Comment (vgao1996) [inline aptos-move/framework/aptos-framework/sources/state_storage.move:31]: This borrow seems unnecessary or is it that you are still working on it?

File: aptos-move/framework/src/natives/state_storage_context.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+/// FIXME(aldenhu): doc
+pub struct StateStorageUsage {
+    items: usize,
+    bytes: usize,
+}
+
+/// FIXME(aldenhu): doc
+/// A table resolver which needs to be provided by the environment. This allows to lookup
+/// data in remote storage, as well as retrieve cost of table operations.
```
Comment (vgao1996) [inline aptos-move/framework/src/natives/state_storage_context.rs:29]: Seems like this is old documentation copied from the table extension impl.

File: aptos-move/framework/src/natives/state_storage_context.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Aptos
```
Comment (vgao1996) [inline aptos-move/framework/src/natives/state_storage_context.rs:1]: Overall this looks good to me.

File: aptos-move/framework/aptos-framework/sources/state_storage.move (hunk 1 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/state_storage.move:31]: yeah already removed locally -- yes, still working on all

File: aptos-move/framework/src/natives/state_storage_context.rs (hunk 2 above)
Comment (msmouse) [reply aptos-move/framework/src/natives/state_storage_context.rs:29]: yeah will fix all the `FIXME(aldenhu)`s before publishing it

**Review by vgao1996 — APPROVED** [review]
LGTM @zekun000 you wanna have a final look?

File: aptos-move/framework/src/natives/state_storage.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+/***************************************************************************************************
+ * native get_state_storage_usage_only_at_eopch_beginning
+ *
+ *   gas cost: base_cost
+ *
+ **************************************************************************************************/
+#[derive(Clone, Debug)]
+pub struct GetUsageGasParameters {
+    pub base_cost: u64,
```
Comment (vgao1996) [inline aptos-move/framework/src/natives/state_storage.rs:46]: @msmouse you need to define an on-chain gas entry for this in the `aptos-gas` crate. See `aptos_framework.rs` and `move_stdlib.rs` for examples.

File: aptos-move/aptos-gas/src/aptos_framework.rs [hunk 5]
```diff
@@ -55,4 +55,6 @@ crate::natives::define_gas_parameters_for_natives!(GasParameters, "aptos_framewo
     [.code.request_publish.unit_cost, "code.request_publish.unit", 1],
 
     [.event.write_to_event_store.unit_cost, "event.write_to_event_store.unit", 1],
+
+    [.state_storage.get_usage.base_cost, "state_storage.get_usage.base", 1]
```
Comment (msmouse) [inline aptos-move/aptos-gas/src/aptos_framework.rs:59]: @vgao1996 this looks okay?

Comment (vgao1996) [reply aptos-move/aptos-gas/src/aptos_framework.rs:59]: Yes

File: aptos-move/framework/aptos-framework/sources/block.move [hunk 6]
```diff
@@ -108,6 +109,10 @@ module aptos_framework::block {
         // transition is the last block in the previous epoch.
         stake::update_performance_statistics(proposer_index_optional, failed_proposer_indices);
 
+        if (round == 0) {
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/block.move:112]: no this doesn't work, we don't have round 0 block. it's even worse, the first round in the new epoch can be different

File: execution/executor-types/src/in_memory_state_calculator.rs [hunk 7]
```diff
@@ -245,39 +266,43 @@ impl InMemoryStateCalculator {
...
+    usage: &mut StateStorageUsage,
     state_key: StateKey,
     write_op: WriteOp,
 ) -> Result<(StateKey, Option<StateValue>)> {
+    let key_size = state_key.size();
     let state_value = match write_op {
         WriteOp::Modification(new_value) | WriteOp::Creation(new_value) => {
-            Some(StateValue::from(new_value))
+            let value = StateValue::from(new_value);
+            usage.items += 1;
```
Comment (zekun000) [inline execution/executor-types/src/in_memory_state_calculator.rs:292]: does the item differentiate Creation vs Modification?

Comment (msmouse) [reply execution/executor-types/src/in_memory_state_calculator.rs:292]: Doesn't matter for calculating total size. We can charge differently before losing that information (when still in the adaptor).

File: aptos-move/framework/aptos-framework/sources/block.move (hunk 6 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/block.move:112]: So how?

Comment (zekun000) [reply aptos-move/framework/aptos-framework/sources/block.move:112]: well you need to maintain something yourself that reset at end of the epoch

Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/block.move]: see the new commit

Comment (zekun000) [reply aptos-move/framework/aptos-framework/sources/block.move]: lgtm

File: execution/executor-types/src/in_memory_state_calculator.rs (hunk 7 above)
Comment (vgao1996) [reply execution/executor-types/src/in_memory_state_calculator.rs:292]: Yes, this information will likely only be used for gas metering in the short term. However other components can follow up should they find the need

**Comment by vgao1996** [comment]
@msmouse It's a bit unfortunate, but you gotta do another rebase and regenerate the e2e-move-tests goldens.

**Pull request merged**